File: GameServer/Proto/LuaProto/Shop_BuyProductProto.lua
--客户端发送购买商城物品消息
Shop_BuyProductProto = { ProtoCode = 16001, ProductId = 0 }

--这句是重定义元表的索引，就是说有了这句，这个才是一个类
Shop_BuyProductProto.__index = Shop_BuyProductProto;

function Shop_BuyProductProto.New()
    local self = { }; --初始化self
    setmetatable(self, Shop_BuyProductProto); --将self的元表设定为Class
    return self;
end


--发送协议
function Shop_BuyProductProto.SendProto(proto)

    local ms = CS.LuaHelper.Instance:CreateMemoryStream();
    ms:WriteUShort(proto.ProtoCode);

    ms:WriteInt(proto.ProductId);

    CS.LuaHelper.Instance:SendProto(ms:ToArray());
    ms:Dispose();
end


--解析协议
function Shop_BuyProductProto.GetProto(buffer)

    local proto = Shop_BuyProductProto.New(); --实例化一个协议对象
    local ms = CS.LuaHelper.Instance:CreateMemoryStream(buffer);

    proto.ProductId = ms:ReadInt();

    ms:Dispose();
    return proto;
end

File: GameServer/Proto/Shop_BuyProductProto.cs

using System.Collections;
using System.Collections.Generic;
using System;

/// <summary>
/// 客户端发送购买商城物品消息
/// </summary>
public struct Shop_BuyProductProto : IProto
{
    public ushort ProtoCode { get { return 16001; } }

    public int ProductId; //商品编号

    public byte[] ToArray()
    {
        using (MMO_MemoryStream ms = new MMO_MemoryStream())
        {
            ms.WriteUShort(ProtoCode);
            ms.WriteInt(ProductId);
            return ms.ToArray();
        }
    }

    public static Shop_BuyProductProto GetProto(byte[] buffer)
    {
        Shop_BuyProductProto proto = new Shop_BuyProductProto();
        using (MMO_MemoryStream ms = new MMO_MemoryStream(buffer))
        {
            proto.ProductId = ms.ReadInt();
        }
        return proto;
    }
}